subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/DirectDependencyMetadataAdapter.java

```diff
@@ -42,4 +48,29 @@ public void doNotEndorseStrictVersions() {
     public boolean isEndorsingStrictVersions() {
         return getOriginalMetadata().isEndorsingStrictVersions();
     }
+
+    @Nullable
+    @Override
+    public String getClassifier() {
+        return getIvyArtifact().map(IvyArtifactName::getClassifier).orElse(null);
+    }
+
+    @Nullable
+    @Override
+    public String getType() {
+        return getIvyArtifact().map(IvyArtifactName::getType).orElse(null);
+    }
+
+    private Optional<IvyArtifactName> getIvyArtifact() {
+        ModuleDependencyMetadata originalMetadata = getOriginalMetadata();
+        if(originalMetadata instanceof ConfigurationBoundExternalDependencyMetadata) {
```

Comment:
I looked at the other implementations (ivy and GMM) and we can also have dependency artifacts there. See also my comment on the tests about the compatibility layer in GMM.

For Gradle Module Metadata it is `GradleDependencyMetadata.getDependencyArtifact()`

For Ivy it is `IvyDependencyDescriptor.getDependencyArtifacts()`

I think it would be good to support these as well for completeness here. In the case of Ivy we have a list, because Ivy allows multiple artifacts on one dependency (https://ant.apache.org/ivy/history/2.1.0/ivyfile/dependency-artifact.html).

Thinking about this, then it would be nice if we could expose all information of `IvyArtifactName` since in ivy you have more flexibility.

Maybe the api could look like this:
```
dep.artifactSelectors // <- returns a list

dep.artifactSelectors.any { it.classifier == 'classy' } // this is how you would check if there is a certain classifier in pom
```

`artifactSelectors` would be of type `List<org.gradle.api.artifacts.DependencyArtifact>` (which is the public equivalent to IvyArtifactName

Reply:
Thanks for the detailled explanation! Reworked the API to expose the `artifactSelectors` as suggested